{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "vzdJq4oPZ_30"
   },
   "source": [
    "# Assignment 2: Building a Simple Index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "9ev2AJMIZ_31"
   },
   "source": [
    "In this assignment, we will build a simple search index, which we will use later for Boolean retrieval. The assignment tasks are again at the bottom of this document."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Edoy7Nw0Z_32"
   },
   "source": [
    "## Loading the Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "RuJ-9QOXZ_32",
    "tags": []
   },
   "outputs": [],
   "source": [
    "Summaries_file = 'data/pandemic_Summaries.pkl.bz2'\n",
    "Abstracts_file = 'data/pandemic_Abstracts.pkl.bz2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 367
    },
    "executionInfo": {
     "elapsed": 9,
     "status": "error",
     "timestamp": 1699642890261,
     "user": {
      "displayName": "Y Wang",
      "userId": "04653720980646316291"
     },
     "user_tz": -60
    },
    "id": "Jy_tFQkeZ_32",
    "outputId": "a8e4bdb9-0464-423b-c2a1-d5ecf76547c2",
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "import pickle, bz2\n",
    "from collections import namedtuple\n",
    "\n",
    "Summaries = pickle.load( bz2.BZ2File( Summaries_file, 'rb' ) )\n",
    "\n",
    "paper = namedtuple( 'paper', ['title', 'authors', 'year', 'doi'] )\n",
    "\n",
    "for (id, paper_info) in Summaries.items():\n",
    "    Summaries[id] = paper( *paper_info )\n",
    "\n",
    "Abstracts = pickle.load( bz2.BZ2File( Abstracts_file, 'rb' ) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "FqOk2FtCZ_33"
   },
   "source": [
    "Let's have a look at what the data looks like for an example of a paper:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "cHKKXP1iZ_33",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "paper(title='Evolution of human influenza A viruses in nature: recombination contributes to genetic variation of H1N1 strains.', authors=['Young JF', 'Palese P'], year=1979, doi='10.1073/pnas.76.12.6547')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Summaries[293742]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "ipoZty81Z_33",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'In June of 1977, a new influenza A pandemic was started by strains of the H1N1 serotype. Oligonucleotide fingerprint analysis of the RNA from viruses isolated during the early stage of this pandemic demonstrated that genetic variation among these 1977 strains could be attributed to sequential mutation [Young, J.F., Desselberger, U. & Palese, P. (1979) Cell, 18, 73-83]. Examination of more recent strains revealed that the H1N1 variants that were isolated in the winter of 1978-1979 differed considerably from the H1N1 viruses isolated the previous year. Oligonucleotide and peptide map analysis of the new prototype strain (A/Cal/10/78) suggested that it arose by recombination. It appears that only the HA, NA, M, and NS genes of this virus are derived from the earlier H1N1 viruses and that the P1, P2, P3, and NP genes most likely originate from an H3N2 parent. These data suggest that genetic variation in influenza virus strains of the same serotype is not restricted to mutation alone, but can also involve recombination (reassortment).'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Abstracts[293742]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "WMPtThheZ_34"
   },
   "source": [
    "## Some Utility Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "0r88Pg3SZ_34"
   },
   "source": [
    "We'll define some utility functions that allow us to tokenize a string into terms, perform linguistic preprocessing on a list of terms, as well as a function to display information about a paper in a nice way. Note that these tokenization and preprocessing functions are rather naive. We will improve them in a later assignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "2MuvRQqoZ_34",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['recognition', 'of', 'the', 'positive', 'effects', 'of', 'music', 'and', 'dance', 'for', 'mental', 'health.']\n"
     ]
    }
   ],
   "source": [
    "def tokenize(text):\n",
    "    \"\"\"\n",
    "    Function that tokenizes a string in a rather naive way. Can be extended later.\n",
    "    \"\"\"\n",
    "    return text.split(' ')\n",
    "\n",
    "def preprocess(tokens):\n",
    "    \"\"\"\n",
    "    Perform linguistic preprocessing on the list of tokens. Can be extended later.\n",
    "    \"\"\"\n",
    "    result = []\n",
    "    for token in tokens:\n",
    "        result.append(token.lower())\n",
    "    return result\n",
    "\n",
    "print(preprocess(tokenize(\"Recognition of the positive effects of music and dance for mental health.\")))\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "ka9YSL4DZ_34",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1073/pnas.76.12.6547>Evolution of human influenza A viruses in nature: recombination contributes to genetic variation of H1N1 strains.</a></strong><br>1979. Young JF, Palese P<br><small><strong>Abstract:</strong> <em>In June of 1977, a new influenza A pandemic was started by strains of the H1N1 serotype. Oligonucleotide fingerprint analysis of the RNA from viruses isolated during the early stage of this pandemic demonstrated that genetic variation among these 1977 strains could be attributed to sequential mutation [Young, J.F., Desselberger, U. & Palese, P. (1979) Cell, 18, 73-83]. Examination of more recent strains revealed that the H1N1 variants that were isolated in the winter of 1978-1979 differed considerably from the H1N1 viruses isolated the previous year. Oligonucleotide and peptide map analysis of the new prototype strain (A/Cal/10/78) suggested that it arose by recombination. It appears that only the HA, NA, M, and NS genes of this virus are derived from the earlier H1N1 viruses and that the P1, P2, P3, and NP genes most likely originate from an H3N2 parent. These data suggest that genetic variation in influenza virus strains of the same serotype is not restricted to mutation alone, but can also involve recombination (reassortment).</em></small><br>[ID: 293742]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "import re\n",
    "\n",
    "def display_summary( id, show_abstract=False, show_id=True, extra_text='' ):\n",
    "    \"\"\"\n",
    "    Function for printing a paper's summary through IPython's Rich Display System.\n",
    "    Trims long author lists, and adds a link to the paper's DOI (when available).\n",
    "    \"\"\"\n",
    "    s = Summaries[id]\n",
    "    lines = []\n",
    "    title = s.title\n",
    "    if s.doi != '':\n",
    "        title = '<a href=http://dx.doi.org/{:s}>{:s}</a>'.format(s.doi, title)\n",
    "    title = '<strong>' + title + '</strong>'\n",
    "    lines.append(title)\n",
    "    authors = ', '.join( s.authors[:20] ) + ('' if len(s.authors) <= 20 else ', ...')\n",
    "    lines.append(str(s.year) + '. ' + authors)\n",
    "    if (show_abstract):\n",
    "        lines.append('<small><strong>Abstract:</strong> <em>{:s}</em></small>'.format(Abstracts[id]))\n",
    "    if (show_id):\n",
    "        lines.append('[ID: {:d}]'.format(id))\n",
    "    if (extra_text != ''):\n",
    "         lines.append(extra_text)\n",
    "    display( HTML('<br>'.join(lines)) )\n",
    "\n",
    "display_summary(293742, show_abstract=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "xAgeWYIqZ_35"
   },
   "source": [
    "## Creating our first index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L9rXngQsZ_35"
   },
   "source": [
    "We will now create an _inverted index_ based on the words in the titles and abstracts of the papers in our dataset. We will implement our inverted index as a Python dictionary with term strings as keys and posting lists (implemented as Python lists) as values. We include all the tokens we can find in the title and (if available) in the abstract:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "MRyxPdLWZ_35",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "inverted_index = defaultdict(list)\n",
    "\n",
    "# This can take a minute:\n",
    "for id in sorted(Summaries.keys()):\n",
    "    term_set = set(preprocess(tokenize(Summaries[id].title)))\n",
    "    if id in Abstracts:\n",
    "        term_set.update(preprocess(tokenize(Abstracts[id])))\n",
    "    for term in term_set:\n",
    "        inverted_index[term].append(id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "K5Mval9kZ_35"
   },
   "source": [
    "Let's see what's in the index for the example term 'ivermectin':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "yM6FiOtFZ_35",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[32293834, 32462282, 32611256, 32752944, 32863755, 33246694, 33721651, 33748818, 34395311, 34428682, 35545850, 35857778, 36068517, 36214394, 36374569, 37692571, 38371720, 38844670, 39070952, 39458594, 39672345, 39892677, 40598561]\n"
     ]
    }
   ],
   "source": [
    "print(inverted_index['ivermectin'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "QBbUIJXEZ_36"
   },
   "source": [
    "We can now use this inverted index to answer simple one-word queries, for example to show all papers that contain the word 'ivermectin':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "id": "e7l3RXcLZ_36",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.3906/sag-2004-145>Antiviral treatment of COVID-19.</a></strong><br>2020. Şimşek Yavuz S, Ünal S<br><small><strong>Abstract:</strong> <em>Currently, there is not any specific effective antiviral treatment for COVID-19. Although most of the COVID-19 patients have mild or moderate courses, up to 5%–10% can have severe, potentially life threatening course, there is an urgent need for effective drugs. Optimized supportive care remains the mainstay of therapy. There have been more than 300 clinical trials going on, various antiviral and immunomodulating agents are in various stages of evaluation for COVID-19 in those trials and some of them will be published in the next couple of months. Despite the urgent need to find an effective antiviral treatment for COVID-19 through randomized controlled studies, certain agents are being used all over the world based on either in-vitro or extrapolated evidence or observational studies. The most frequently used agents both in Turkey and all over the world including chloroquine, hydroxychloroquine, lopinavir/ritonavir, favipiravir and remdesivir will be reviewed here .Nitazoxanide and ivermectin were also included in this review as they have recently been reported to have an activity against SARS-CoV-2 in vitro and are licensed for the treatment of some other human infections.</em></small><br>[ID: 32293834]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1007/s00210-020-01902-5>Ivermectin, antiviral properties and COVID-19: a possible new mechanism of action.</a></strong><br>2020. Rizzo E<br><small><strong>Abstract:</strong> <em>Ivermectin is an antiparasitic drug that has shown also an effective pharmacological activity towards various infective agents, including viruses. This paper proposes an alternative mechanism of action for this drug that makes it capable of having an antiviral action, also against the novel coronavirus, in addition to the processes already reported in literature.</em></small><br>[ID: 32462282]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1080/10428194.2020.1786559>Continuous high-dose ivermectin appears to be safe in patients with acute myelogenous leukemia and could inform clinical repurposing for COVID-19 infection.</a></strong><br>2020. de Castro CG Jr, Gregianin LJ, Burger JA<br><small><strong>Abstract:</strong> <em></em></small><br>[ID: 32611256]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1080/07391102.2020.1802345>Current status and strategic possibilities on potential use of combinational drug therapy against COVID-19 caused by SARS-CoV-2.</a></strong><br>2021. Siddiqui AJ, Jahan S, Ashraf SA, Alreshidi M, Ashraf MS, Patel M, Snoussi M, Singh R, Adnan M<br><small><strong>Abstract:</strong> <em>The spread of new coronavirus infection starting December 2019 as novel SARS-CoV-2, identified as the causing agent of COVID-19, has affected all over the world and been declared as pandemic. Approximately, more than 8,807,398 confirmed cases of COVID-19 infection and 464,483 deaths have been reported globally till the end of 21 June 2020. Until now, there is no specific drug therapy or vaccine available for the treatment of COVID-19. However, some potential antimalarial drugs like hydroxychloroquine and azithromycin, antifilarial drug ivermectin and antiviral drugs have been tested by many research groups worldwide for their possible effect against the COVID-19. Hydroxychloroquine and ivermectin have been identified to act by creating the acidic condition in cells and inhibiting the importin (IMPα/β1) mediated viral import. There is a possibility that some other antimalarial drugs/antibiotics in combination with immunomodulators may help in combatting this pandemic disease. Therefore, this review focuses on the current use of various drugs as single agents (hydroxychloroquine, ivermectin, azithromycin, favipiravir, remdesivir, umifenovir, teicoplanin, nitazoxanide, doxycycline, and dexamethasone) or in combinations with immunomodulators additionally. Furthermore, possible mode of action, efficacy and current stage of clinical trials of various drug combinations against COVID-19 disease has also been discussed in detail.Communicated by Ramaswamy H. Sarma.</em></small><br>[ID: 32752944]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.21315/mjms2020.27.4.15>Ivermectin: Potential Role as Repurposed Drug for COVID-19.</a></strong><br>2020. Dixit A, Yadav R, Singh AV<br><small><strong>Abstract:</strong> <em>Severe acute respiratory illness caused by 2019 novel coronavirus (2019-nCoV), officially named severe acute respiratory syndrome coronavirus (SARS-CoV-2) in late December 2019 is an extremely communicable disease. World Health Organization (WHO) declared coronavirus disease 2019 (COVID-19) as a pandemic as it has spread to at least 200 countries in a short span of time. Being a new disease there is lack of information about pathogenesis and proliferation pathways of this new coronavirus. Currently there is no effective treatment for coronavirus infection; major effort is to develop vaccine against the virus and development of therapeutic drugs for the disease. The development of genome-based vaccine and therapeutic antibodies require thorough testing for safety and will be available after some time. In the meanwhile, the available practical approach is to repurpose existing therapeutic agents, with proven safety record as a rapid response measure for the current pandemic. Here we discuss the presently used repurposed drugs for COVID-19 and the potential for ivermectin (IVM) to be used as a therapeutic option in COVID-19.</em></small><br>[ID: 32863755]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1016/j.mehy.2020.110364>Inhaled route and anti-inflammatory action of ivermectin: Do they hold promise in fighting against COVID-19?</a></strong><br>2021. Mittal N, Mittal R<br><small><strong>Abstract:</strong> <em>In an effort to curb the global pandemic due to coronavirus, the scientific community is exploring various treatment strategies with a special emphasis on drug repurposing. Ivermectin, an anti-helminthic drug is also being proposed for treatment and prevention of COVID-19. Ivermectin has demonstrated broad spectrum antiviral activity against both DNA and RNA viruses. Due to its potential to interfere with transport of SARS-CoV-2 nucleocapsid protein to nucleus, it is being proposed to have antiviral activity against this virus as well which has been confirmed in an in-vitro study. However, in-vitro to in-vivo extrapolation studies indicate an inability to achieve the desired IC50 levels of ivermectin after oral administration of doses up to 10 times higher than the approved anti-helminthic dose. In a modelling simulation study, drug accumulation in the lungs was noticed at levels having potential antiviral activity. It is hypothesised that inhaled formulation of ivermectin may be effective against SARS-CoV-2. Therefore, ivermectin administered via inhalational route needs to be explored for potential beneficial role in COVID-19 in preclinical and clinical studies. We also hypothesise the possibility of drug having anti-inflammatory action in coronavirus associated severe respiratory illness based on few in-vitro and in-vivo reports which however needs to be confirmed clinically.</em></small><br>[ID: 33246694]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1016/j.scitotenv.2021.146257>Environmental impact assessment of COVID-19 therapeutic solutions. A prospective analysis.</a></strong><br>2021. Tarazona JV, Martínez M, Martínez MA, Anadón A<br><small><strong>Abstract:</strong> <em>Several medicinal products for human use are currently under consideration as potential treatment for COVID-19 pandemic. As proposals cover also prophylactic use, the treatment could be massive, resulting in unprecedent levels of antiviral emissions to the aquatic environment. We have adapted previous models and used available information for predicting the environmental impact of representative medicinal products, covering the main groups under consideration: multitarget antiparasitic (chloroquines and ivermectin), glucocorticoids, macrolide antibiotics and antiviral drugs including their pharmacokinetic boosters. The retrieved information has been sufficient for conducting a conventional environmental risk assessment for the group of miscellaneous medicines; results suggest low concern for the chloroquines and dexamethasone while very high impact for ivermectin and azithromycin, even at use levels well below the default value of 1% of the population. The information on the ecotoxicity of the antiviral medicines is very scarce, thus we have explored an innovative pharmacodynamic-based approach, combining read-across, quantitative structure-activity relationship (QSAR), US EPA's Toxicity Forecaster (ToxCast) in vitro data, pharmacological modes of action, and the observed adverse effects. The results highlight fish sublethal effects as the most sensitive target and identify possible concerns. These results offer guidance for minimizing the environmental risk of treatment medication for COVID-19.</em></small><br>[ID: 33721651]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1055/s-0041-1723086>Potential of CRISPR/Cas13 System in Treatment and Diagnosis of COVID-19.</a></strong><br>2021. Khodavirdipour A, Piri M, Jabbari S, Khalaj-Kondori M<br><small><strong>Abstract:</strong> <em>The novel coronavirus disease 2019 (COVID-19) belongs to coronaviridae families like sarbecovirus (SARS), and causes pyrexia, pertussis, and acute respiratory distress syndrome (ARDS) in major. Started from Wuhan, China, COVID-19 now forced the World Health Organization (WHO) call it a global pandemic. These dreadful figures elevate the need for rapid action for a rapid diagnostic tool, an efficacious therapy, or vaccine for such widespread disease. In this article, we reviewed all the latest research and trials including conventional antiviral medicines that have a narrow and finite effect on COVID-19. Recently, some advances were made by a nucleotide/nucleoside analogues (NUC) inhibitor (remdesivir), ivermectin (antiparasitic drug), and convalescent plasma; the later one has more recently been approved by the Food and Drug Administration (FDA). Additionally, a clinical-grade soluble human angiotensin-converting enzyme (ACE2), named hrsACE2, was able to inhibit the infection of human blood vessel organoids, as well as the human kidney organoids, by the virus. As of now, innovative therapeutics based on the CRISPR/Cas13d might overcome the challenge of COVID-19 either as a treatment option or precise and rapid diagnostic tool due to its rapid and precise nature. In this updated comprehensive rapid review, we tried to cover all recent findings in terms of genomics, diagnosis, prevention, and treatment.</em></small><br>[ID: 33748818]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.3389/fcimb.2021.700502>Combination Treatment With Remdesivir and Ivermectin Exerts Highly Synergistic and Potent Antiviral Activity Against Murine Coronavirus Infection.</a></strong><br>2021. Tan YL, Tan KSW, Chu JJH, Chow VT<br><small><strong>Abstract:</strong> <em>The recent COVID-19 pandemic has highlighted the urgency to develop effective antiviral therapies against the disease. Murine hepatitis virus (MHV) is a coronavirus that infects mice and shares some sequence identity to SARS-CoV-2. Both viruses belong to the </em></small><br>[ID: 34395311]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1016/j.bpc.2021.106677>Comparative study of the interaction of ivermectin with proteins of interest associated with SARS-CoV-2: A computational and biophysical approach.</a></strong><br>2021. González-Paz L, Hurtado-León ML, Lossada C, Fernández-Materán FV, Vera-Villalobos J, Loroño M, Paz JL, Jeffreys L, Alvarado YJ<br><small><strong>Abstract:</strong> <em>The SARS-CoV-2 pandemic has accelerated the study of existing drugs. The mixture of homologs called ivermectin (avermectin-B1a [HB1a] + avermectin-B1b [HB1b]) has shown antiviral activity against SARS-CoV-2 in vitro. However, there are few reports on the behavior of each homolog. We investigated the interaction of each homolog with promising targets of interest associated with SARS-CoV-2 infection from a biophysical and computational-chemistry perspective using docking and molecular dynamics. We observed a differential behavior for each homolog, with an affinity of HB1b for viral structures, and of HB1a for host structures considered. The induced disturbances were differential and influenced by the hydrophobicity of each homolog and of the binding pockets. We present the first comparative analysis of the potential theoretical inhibitory effect of both avermectins on biomolecules associated with COVID-19, and suggest that ivermectin through its homologs, has a multiobjective behavior.</em></small><br>[ID: 34428682]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1111/jfbc.14219>In silico evaluation of food-derived carotenoids against SARS-CoV-2 drug targets: Crocin is a promising dietary supplement candidate for COVID-19.</a></strong><br>2022. Mujwar S, Sun L, Fidan O<br><small><strong>Abstract:</strong> <em>The current COVID-19 pandemic is severely threatening public healthcare systems around the globe. Some supporting therapies such as remdesivir, favipiravir, and ivermectin are still under the process of a clinical trial, it is thus urgent to find alternative treatment and prevention options for SARS-CoV-2. In this regard, although many natural products have been tested and/or suggested for the treatment and prophylaxis of COVID-19, carotenoids as an important class of natural products were underexplored. The dietary supplementation of some carotenoids was already suggested to be potentially effective in the treatment of COVID-19 due to their strong antioxidant properties. In this study, we performed an in silico screening of common food-derived carotenoids against druggable target proteins of SARS-CoV-2 including main protease, helicase, replication complex, spike protein and its mutants for the recent variants of concern, and ADP-ribose phosphatase. Molecular docking results revealed that some of the carotenoids had low binding energies toward multiple receptors. Particularly, crocin had the strongest binding affinity (-10.5 kcal/mol) toward the replication complex of SARS-CoV-2 and indeed possessed quite low binding energy scores for other targets as well. The stability of crocin in the corresponding receptors was confirmed by molecular dynamics simulations. Our study, therefore, suggests that carotenoids, especially crocin, can be considered an effective alternative therapeutics and a dietary supplement candidate for the prophylaxis and treatment of SARS-CoV-2. PRACTICAL APPLICATIONS: In this study, food-derived carotenoids as dietary supplements have the potential to be used for the prophylaxis and/or treatment of SARS-CoV-2. Using in silico techniques, we aimed at discovering food-derived carotenoids with inhibitory effects against multiple druggable sites of SARS-CoV-2. Molecular docking experiments against main protease, helicase, replication complex, spike protein and its mutants for the recent variants of concern, and ADP-ribose phosphatase resulted in a few carotenoids with multitarget inhibitory effects. Particularly, crocin as one of the main components of saffron exhibited strong binding affinities to the multiple drug targets including main protease, helicase, replication complex, mutant spike protein of lineage B.1.351, and ADP-ribose phosphatase. The stability of the crocin complexed with these drug targets was further confirmed through molecular dynamics simulations. Overall, our study provides the preliminary data for the potential use of food-derived carotenoids, particularly crocin, as dietary supplements in the prevention and treatment of COVID-19.</em></small><br>[ID: 35545850]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1371/journal.pntd.0010044>Step towards elimination of Wuchereria bancrofti in Southwest Tanzania 10 years after mass drug administration with Albendazole and Ivermectin.</a></strong><br>2022. Mnkai J, Marandu TF, Mhidze J, Urio A, Maganga L, Haule A, Kavishe G, Ntapara E, Chiwerengo N, Clowes P, Horn S, Mosoba M, Lazarus W, Ngenya A, Kalinga A, Debrah A, Rieß F, Saathoff E, Geldmacher C, Hoerauf A, ...<br><small><strong>Abstract:</strong> <em>Lymphatic filariasis is a mosquito transmitted parasitic infection in tropical regions. Annual mass treatment with ivermectin and albendazole is used for transmission control of Wuchereria bancrofti, the infective agent of lymphatic filariasis in many African countries, including Tanzania.</em></small><br>[ID: 35857778]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1186/s12916-022-02486-y>The prevalence of onchocerciasis in Africa and Yemen, 2000-2018: a geospatial analysis.</a></strong><br>2022. Schmidt CA, Cromwell EA, Hill E, Donkers KM, Schipp MF, Johnson KB, Pigott DM, LBD 2019 Neglected Tropical Diseases Collaborators, Hay SI<br><small><strong>Abstract:</strong> <em>Onchocerciasis is a disease caused by infection with Onchocerca volvulus, which is transmitted to humans via the bite of several species of black fly, and is responsible for permanent blindness or vision loss, as well as severe skin disease. Predominantly endemic in parts of Africa and Yemen, preventive chemotherapy with mass drug administration of ivermectin is the primary intervention recommended for the elimination of its transmission.</em></small><br>[ID: 36068517]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.7196/SAMJ.2022.v112i8.16473>Ivermectin exposures reported to the Poisons Information Helpline in South Africa during the COVID-19 pandemic.</a></strong><br>2022. Pillay-Fuentes Lorente V, Voigt G, Du Plessis CE, Balme K, Marks CJ, Decloedt EH, Stephen C, Reuter H, Van Rensburg R<br><small><strong>Abstract:</strong> <em>Ivermectin is an antiparasitic drug that has shown in vitro activity against COVID‑19. Clinical studies supporting ivermectin for COVID‑19 prevention and treatment are conflicting, with important limitations. Public support for ivermectin is significant, with extensive off-label use despite the conflicting views on its efficacy. Ivermectin tablets and injectable formulations are not registered in South Africa for human use by the South African Health Products Regulatory Authority. The National Department of Health does not currently recommend the use of ivermectin for COVID‑19.</em></small><br>[ID: 36214394]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1093/jac/dkac382>Outpatient purchasing patterns of hydroxychloroquine and ivermectin in the USA and Canada during the COVID-19 pandemic: an interrupted time series analysis from 2016 to 2021.</a></strong><br>2022. Ho M, Tadrous M, Iacono A, Suda K, Gomes T<br><small><strong>Abstract:</strong> <em>Hydroxychloroquine and ivermectin received widespread attention after initial studies suggested that they were effective against COVID-19. However, several of these studies were later discredited.</em></small><br>[ID: 36374569]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.7759/cureus.43168>COVID-19 Excess Deaths in Peru's 25 States in 2020: Nationwide Trends, Confounding Factors, and Correlations With the Extent of Ivermectin Treatment by State.</a></strong><br>2023. Chamie JJ, Hibberd JA, Scheim DE<br><small><strong>Abstract:</strong> <em>Introduction In 2020, nations hastened to contain an emerging COVID-19 pandemic by deploying diverse public health approaches, but conclusive appraisals of the efficacy of these approaches are elusive in most cases. One of the medicines deployed, ivermectin (IVM), a macrocyclic lactone having biochemical activity against SARS-CoV-2 through competitive binding to its spike protein, has yielded mixed results in randomized clinical trials (RCTs) for COVID-19 treatments. In Peru, an opportunity to track the efficacy of IVM with a close consideration of confounding factors was provided through data for excess deaths as correlated with IVM use in 2020, under semi-autonomous policies in its 25 states. Methods To evaluate possible IVM treatment effects, excess deaths as determined from Peruvian national health data were analyzed by state for ages ≥60 in Peru's 25 states. These data were compared with monthly summary data for excess deaths in Peru for the period 2020-2021 as published by the WHO in 2022. To identify potential confounding factors, Google mobility data, population densities, SARS-CoV-2 genetic variations, and seropositivity rates were also examined. Results Reductions in excess deaths over a period of 30 days after peak deaths averaged 74% in the 10 states with the most intensive IVM use. As determined across all 25 states, these reductions in excess deaths correlated closely with the extent of IVM use (p<0.002). During four months of IVM use in 2020, before a new president of Peru restricted its use, there was a 14-fold reduction in nationwide excess deaths and then a 13-fold increase in the two months following the restriction of IVM use. Notably, these trends in nationwide excess deaths align with WHO summary data for the same period in Peru. Conclusions The natural experiment that was put into motion with the authorization of IVM use for COVID-19 in Peru in May 2020, as analyzed using data on excess deaths by locality and by state from Peruvian national health sources, resulted in strong evidence for the drug's effectiveness. Several potential confounding factors, including effects of a social isolation mandate imposed in May 2020, variations in the genetic makeup of the SARS-CoV-2 virus, and differences in seropositivity rates and population densities across the 25 states, were considered but did not appear to have significantly influenced these outcomes.</em></small><br>[ID: 37692571]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.51866/oa.415>Knowledge, attitude and prevalence of ivermectin use as coronavirus disease treatment: A crosssectional study among a Malaysian population.</a></strong><br>2024. Filza Nur Athirah T, Aina Amanina AJ, Nor Safwan Hadi NA<br><small><strong>Abstract:</strong> <em>During the coronavirus disease (COVID-19) pandemic, there was a widespread public misconception regarding ivermectin use in managing the disease. There is no approval of ivermectin as COVID-19 treatment by the Food and Drug Administration, and the adverse effects of the drug are alarming. This study aimed to evaluate the knowledge and attitude towards and prevalence of ivermectin use as COVID-19 treatment among the Malaysian population.</em></small><br>[ID: 38371720]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1007/s00403-024-03004-1>Post Covid telogen effluvium: the diagnostic value of serum ferritin biomarker and the preventive value of dietary supplements. a case control study.</a></strong><br>2024. Bedair NI, Abdelaziz AS, Abdelrazik FS, El-Kassas M, AbouHadeed MH<br><small><strong>Abstract:</strong> <em>Telogen effluvium is characterized by excessive hair shedding usually following a stressful event. Ferritin has been used in clinical practice as a biomarker of nonanemic iron deficiency in cases of telogen effluvium. During the years of the COVID19 pandemic, telogen effluvium was reported as a part of post covid manifestations. As ferritin was also a biomarker for inflammation in cases with covid infection, this study was designed to evaluate the value of ferritin in cases with postcovid telogen effluvium one hundred patients recovering from covid 19 for 4-12 weeks were included in the study, detailed drug and laboratory history was obtained and serum ferritin level was measured. the mean serum level of ferritin among telogen effluvium patients was significantly lower than controls (68.52 ± 126 and 137 ± 137.597 ug/L respectively). Patients with telogen effluvium used significantly more azithromycin and ivermectin and significantly less vitamin C, D, lactoferrin and zinc than the controls Although serum ferritin is lower among telogen effluvium patients, it was still higher than the cutoff value for diagnosing nonanemic iron deficiency, we suggest that it will not be a good biomarkers in these cases. Our secondary outcomes showed that dietary supplements used during active infection such as vitamin C, D, lactoferrin and zinc might have a preventive value on postcovid hair loss, while azithromycin and ivermectin could have a negative long term effect on telogen effluvium.</em></small><br>[ID: 38844670]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1177/2632010X241263054>Preclinical and Clinical Investigations of Potential Drugs and Vaccines for COVID-19 Therapy: A Comprehensive Review With Recent Update.</a></strong><br>2024. Mia ME, Howlader M, Akter F, Hossain MM<br><small><strong>Abstract:</strong> <em>The COVID-19 pandemic-led worldwide healthcare crisis necessitates prompt societal, ecological, and medical efforts to stop or reduce the rising number of fatalities. Numerous mRNA based vaccines and vaccines for viral vectors have been licensed for use in emergencies which showed 90% to 95% efficacy in preventing SARS-CoV-2 infection. However, safety issues, vaccine reluctance, and skepticism remain major concerns for making mass vaccination a successful approach to treat COVID-19. Hence, alternative therapeutics is needed for eradicating the global burden of COVID-19 from developed and low-resource countries. Repurposing current medications and drug candidates could be a more viable option for treating SARS-CoV-2 as these therapies have previously passed a number of significant checkpoints for drug development and patient care. Besides vaccines, this review focused on the potential usage of alternative therapeutic agents including antiviral, antiparasitic, and antibacterial drugs, protease inhibitors, neuraminidase inhibitors, and monoclonal antibodies that are currently undergoing preclinical and clinical investigations to assess their effectiveness and safety in the treatment of COVID-19. Among the repurposed drugs, remdesivir is considered as the most promising agent, while favipiravir, molnupiravir, paxlovid, and lopinavir/ritonavir exhibited improved therapeutic effects in terms of elimination of viruses. However, the outcomes of treatment with oseltamivir, umifenovir, disulfiram, teicoplanin, and ivermectin were not significant. It is noteworthy that combining multiple drugs as therapy showcases impressive effectiveness in managing individuals with COVID-19. Tocilizumab is presently employed for the treatment of patients who exhibit COVID-19-related pneumonia. Numerous antiviral drugs such as galidesivir, griffithsin, and thapsigargin are under clinical trials which could be promising for treating COVID-19 individuals with severe symptoms. Supportive treatment for patients of COVID-19 may involve the use of corticosteroids, convalescent plasma, stem cells, pooled antibodies, vitamins, and natural substances. This study provides an updated progress in SARS-CoV-2 medications and a crucial guide for inventing novel interventions against COVID-19.</em></small><br>[ID: 39070952]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.3390/pharmaceutics16101262>Multidrug Combinations against SARS-CoV-2 Using GS-441524 or Ivermectin with Molnupiravir and/or Nirmatrelvir in Reconstituted Human Nasal Airway Epithelia.</a></strong><br>2024. Siegrist D, Jonsdottir HR, Bouveret M, Boda B, Constant S, Engler OB<br><small><strong>Abstract:</strong> <em> Background. The emergence, global spread, and persistence of SARS-CoV-2 resulted in an unprecedented need for effective antiviral drugs. Throughout the pandemic, various drug development and treatment strategies were adopted, including repurposing of antivirals designed for other viruses along with a multitude of other drugs with varying mechanisms of action (MoAs). Furthermore, multidrug treatment against COVID-19 is an ongoing topic and merits further investigation. Method/Objectives. We assessed the efficacy of multidrug treatment against SARS-CoV-2 in reconstituted human nasal epithelia, using combinations of molnupiravir and nirmatrelvir as a baseline, adding suboptimal concentrations of either GS-441524 or ivermectin, attempting to increase overall antiviral activity while lowering the overall therapeutic dose. Results. Nirmatrelvir combined with molnupiravir, GS-441524, or ivermectin at suboptimal concentrations show increased antiviral activity compared to single treatment. No triple combinations showed improved inhibition of SARS-CoV-2 replication beyond what was observed for double treatments. Conclusions. In general, we observed that the addition of a third compound is not beneficial for antiviral activity, while various double combinations exhibit increased antiviral activity over single treatment. </em></small><br>[ID: 39458594]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1016/j.chemosphere.2024.143933>Rivers at risks: The interplay of \"COVID kit\" medication misuse and urban waterway contaminants.</a></strong><br>2025. Ribeiro Trisotto LF, Figueredo CC, Gomes MP<br><small><strong>Abstract:</strong> <em>This study investigates the environmental impact of the widespread use of \"COVID Kit\" drugs-azithromycin (AZI), ivermectin (IVE), and hydroxychloroquine (HCQ)-in urban rivers of Curitiba in Brazil, during and after the COVID-19 pandemic. The research focuses on the occurrence and concentrations of these pharmaceuticals in water and sediment samples collected from key urban rivers. Concentrations of AZI, IVE, and HCQ in water ranged from 326 to 3340 ng/L, 130-3340 ng/L, and 304-3314 ng/L, respectively, while in sediment, they ranged from 18 to 249 ng/g, 21-480 ng/g, and 38-673 ng/g, respectively. Results indicate a significant increase in AZI, IVE, and HCQ concentrations during the pandemic. Concentrations of these drugs peaked in September 2020 and March 2021, declining after the start of Brazil's vaccination campaign. However, the levels of these pharmaceuticals remained elevated in some areas even after the decline in their usage. Environmental risk assessments were conducted to evaluate the potential ecological hazards posed by these pharmaceuticals, revealing the long-term persistence of these drugs in aquatic environments and their potential to contribute to antimicrobial resistance. The findings of this study underscore the critical need for robust regulatory measures and improved wastewater treatment processes to prevent pharmaceutical contamination in urban water systems, particularly during global health crises.</em></small><br>[ID: 39672345]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1016/j.ijpharm.2025.125302>Inhalable spray-dried dry powders combining ivermectin and niclosamide to inhibit SARS-CoV-2 infection in vitro.</a></strong><br>2025. Su C, Saha T, Sinha S, Hird CP, Smith SXY, Quiñones-Mateu ME, Das SC<br><small><strong>Abstract:</strong> <em>SARS-CoV-2, the virus responsible for the COVID-19 pandemic, predominantly affects the respiratory tract, underscoring the need to develop antiviral agents in an inhalable formulation that can be delivered as prophylactic and/or therapeutic drugs directly to the infection site. Since the beginning of the pandemic, our group has been exploring the possibility of developing combinations of antiviral drugs that can be delivered as inhalable therapy, including combinations of remdesivir and ebselen or remdesivir and disulfiram prepared using a spray-drying technique. In this study, we used a similar spray-drying technique to develop inhalable dry powders combining the controversial drugs ivermectin and niclosamide, which have been reported to exhibit synergistic activity against SARS-CoV-2 in vitro. The combined dry powders were within the size range of 1-5 μm, amorphous in nature and displayed characteristic morphology after spray drying. The emitted dose (ED) of the spray-dried powders ranged from 68 to 83 %, whereas the fine particle fraction (FPF) ranged between 50 and 74 %. All the prepared dry powders remained stable under different humidity conditions (<15 % RH and 53 % RH). Interestingly, the optimized combinational dry powder of ivermectin and niclosamide showed an improved cytotoxic profile (CC</em></small><br>[ID: 39892677]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1186/s13071-025-06837-5>Simulium larvae susceptibility to temephos and the effect of 10 weeks of treatment of the Mbende tributary in the Nkam-Wouri River drainage of Cameroon on larval density and adult fly biting rates.</a></strong><br>2025. Abong RA, Ekanya R, Nji TM, Fombad FF, Amambo GN, Ngong I, Nietcho FN, Deribe K, Biholong B, Njiokou F, Ekobo S, Wondji C, Enyong P, Wanji S<br><small><strong>Abstract:</strong> <em>Despite over 18 years of annual ivermectin mass drug administration (MDA) in Cameroon's Nkam-Wouri River drainage, onchocerciasis transmission persists. Several reasons, including multiple breeding sites and abundant vector populations, contribute to ongoing transmission. High vector abundance also causes a biting nuisance to local populations. The change in paradigm from onchocerciasis control to elimination may not be achieved if alternative control measures are not used. There is a need to complement ivermectin MDA with other strategies. This study tested the susceptibility of Simulium larvae to temephos insecticide and monitored the effect of 10 weeks of ground larviciding on the larval density and black fly population.</em></small><br>[ID: 40598561]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query_word = 'ivermectin'\n",
    "for i in inverted_index[query_word]:\n",
    "    display_summary(i, show_abstract=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "cLS3Rm9iZ_36"
   },
   "source": [
    "----------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "GVhdywUWZ_36"
   },
   "source": [
    "# Tasks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "xjEBsuRmZ_36"
   },
   "source": [
    "**Your name:** ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jSVisMaSZ_36"
   },
   "source": [
    "### Task 1\n",
    "\n",
    "Implement the function `sorted_merge_and` outlined below. This function takes two posting lists from the index that can be assumed to be sorted already (e.g. [3,5,8,11,18] and [5,7,8,9,11,12]), and it should return the result of the merging of the two lists with AND. The resulting list should therefore include all the elements that appear in both lists. As explained on the slides, this operation should take advantage of the input lists being sorted already, should not perform any additional sorting operation, and should go through each of the input lists just once. Then, test your function with an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "e3zR_ex_Z_37"
   },
   "source": [
    "### Task 2\n",
    "\n",
    "Similarly as above, implement the function `sorted_or` outlined below that executes an OR merging of the lists. The resulting list should therefore include all the elements that appear in at least one of the lists. Again, this operation should take advantage of the input lists being sorted already, should not perform any additional sorting operation, and should go through each of the input lists just once. Elements that appear in both input list should only appear once in the output list. Test your function again with an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "0msSbXQ1Z_37"
   },
   "source": [
    "### Task 3\n",
    "\n",
    "Construct a function called `and_query` that takes as input a single string, consisting of one or more words, and returns as function value a list of matching documents. `and_query`, as its name suggests, should require that all query terms are present in the documents of the result list.\n",
    "\n",
    "For that, access the variable `inverted_index` from above and use the method `merge_and` that you defined. Also use the `tokenize` and `preprocess` functions we defined above to tokenize and preprocess your query.\n",
    "\n",
    "Again demonstrate the working of your function with an example (choose one that leads to fewer than 100 hits to not overblow this notebook file)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "4VRzOafAZ_37"
   },
   "source": [
    "### Task 4\n",
    "\n",
    "Construct another function called `query_or` that works in the same way as `query_and` you just implemented, but returns as function value the documents that contain _at least one_ of the words in the query, using the `sorted_merge_or` function you defined.\n",
    "\n",
    "Demonstrate the working of this function also with an example (again, choose one that leads to fewer than 100 hits)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "HaNZGq8PZ_38"
   },
   "source": [
    "### Task 5\n",
    "\n",
    "Why does `query_and('ivermectin vitro')` not return paper '34428682' and '39892677', even though it mentions these terms in the title and abstract? (You do not have to implement anything to fix this yet!)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "F_vmfd1fZ_38"
   },
   "source": [
    "**Answer:** "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "_vSIYm2rZ_38"
   },
   "source": [
    "# Submission"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "kNrQUUY8Z_38"
   },
   "source": [
    "Submit the answers to the assignment via Canvas as a modified version of this Notebook file (file with `.ipynb` extension) that includes your code and your answers.\n",
    "\n",
    "Before submitting, restart the kernel and re-run the complete code (**Kernel > Restart & Run All**), and then check whether your assignment code still works as expected.\n",
    "\n",
    "Don't forget to add your name, and remember that the assignments have to be done **individually**, and that code sharing or copying are **strictly forbidden** and will be punished."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "zRtM3zsmZ_38"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
